solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceDirectInputOperator.java

```diff
@@ -0,0 +1,238 @@
+import com.solacesystems.jcsmp.BytesXMLMessage;
+import com.solacesystems.jcsmp.Consumer;
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.Topic;
+import com.datatorrent.api.Context;
+import com.datatorrent.api.InputOperator;
+import com.datatorrent.api.Operator;
+import com.datatorrent.netlet.util.DTThrowable;
+
+public abstract class AbstractSolaceDirectInputOperator<T> extends AbstractSolaceBaseInputOperator<T> implements InputOperator, Operator.ActivationListener<com.datatorrent.api.Context.OperatorContext>
+{
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractSolaceDirectInputOperator.class);
+
+  private final transient AtomicReference<Throwable> throwable;
+  private transient Context.OperatorContext context;
+  protected transient long currentWindowId;
+
+  @NotNull
+  protected String topicName;
+  private transient Topic topic;
+  private transient BytesXMLMessage recentMessage = null;
+
+  private transient List<T> messages = new ArrayList<T>();
+  protected final transient Map<Long, T> currentWindowRecoveryState;
+  protected static final transient int DEFAULT_BUFFER_SIZE = 500;
+
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    arrivedTopicMessagesToProcess = new ArrayBlockingQueue<BytesXMLMessage>(Integer.parseInt(this.unackedMessageLimit));
+
+    this.context = context;
+
+    //setup info for HA and DR at the transport level
+    super.setConnectRetries(this.connectRetries);
+    super.setReconnectRetries(this.reconnectRetries);
+    super.setReapplySubscriptions(true);
+
+    super.setup(context);
+    topic = factory.createTopic(topicName);
+    addSubscription(topic);
+  }
+
+  public AbstractSolaceDirectInputOperator()
+  {
+    throwable = new AtomicReference<Throwable>();
+    currentWindowRecoveryState = Maps.newLinkedHashMap();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    currentWindowId = windowId;
+    super.beginWindow(windowId);
+    if (windowId <= lastCompletedWId) {
+      handleRecovery(currentWindowId);
+    }
+  }
+
+  @Override
+  public void endWindow()
+  {
+    super.endWindow();
+    try {
+      if (windowId > lastCompletedWId) {
+        idempotentStorageManager.save(currentWindowRecoveryState, operatorId, windowId);
+      }
+      currentWindowRecoveryState.clear();
+    } catch (IOException e) {
+      DTThrowable.rethrow(e);
+    }
+    emitCount = 0; //reset emit count
+  }
+
+  @Override
+  public void emitTuples()
+  {
+    if (currentWindowId <= idempotentStorageManager.getLargestRecoveryWindow()) {
+      return;
+    }
+
+    BytesXMLMessage message;
+    try {
+      while (emitCount < DEFAULT_BUFFER_SIZE && ((message = (BytesXMLMessage)super.arrivedTopicMessagesToProcess.poll(10, TimeUnit.MILLISECONDS)) != null)) {
+        processMessage(message);
+        emitCount++;
+      }
+    } catch (InterruptedException e) {
+      DTThrowable.rethrow(e);
+    }
+    emitCount++;
```

Comment:
Why this extra increment of emitCount? Is this correct? If yes, can we have a comment about why this is required?